corehq/motech/openmrs/finders.py

```diff
@@ -0,0 +1,287 @@
+            self._property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for addr_prop, value_source in case_config['person_preferred_address'].items():
+            jsonpath = parse('person.preferredAddress.' + addr_prop)
+            self._property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+        for id_type_uuid, value_source in case_config['patient_identifiers'].items():
+            if id_type_uuid == 'uuid':
+                jsonpath = parse('uuid')
+            else:
+                # "(identifiers[*] where identifierType.uuid=id_type_uuid).identifier"
+                jsonpath = Child(
+                    WhereCmp(
+                        Child(Fields('identifiers'), Slice()),
+                        Child(Fields('identifierType'), Fields('uuid')), eq, id_type_uuid
+                    ),
+                    Fields('identifier')
+                )
+            self._property_map.update(get_caseproperty_jsonpathvaluemap(jsonpath, value_source))
+
+    def get_score(self, patient, case):
+        """
+        Return the sum of weighted properties to give an OpenMRS
+        patient a score of how well they match a CommCare case.
+        """
+        def weights():
+            for property_weight in self.property_weights:
+                prop = property_weight['case_property']
+                weight = property_weight['weight']
+
+                matches = self._property_map[prop].jsonpath.find(patient)
+                if matches:
+                    assert len(matches) == 1, 'jsonpath "{}" did not uniquely match a patient value'.format(
+                        self._property_map[prop].jsonpath
+                    )
+                    patient_value = matches[0].value
+                    value_map = self._property_map[prop].value_map
+                    case_value = case.get_case_property(prop)
+                    is_equal = value_map.get(patient_value, patient_value) == case_value
+                    yield weight if is_equal else 0
+                else:
+                    yield 0
+
+        return sum(weights())
+
+    def save_match_id(self, case, case_config, patient):
+        """
+        If we are confident of the patient matched to a case, save
+        the patient's ID to the case.
+        """
+        from casexml.apps.case.mock import CaseBlock
+        from corehq.apps.hqcase.utils import submit_case_blocks
+        from corehq.motech.openmrs.repeater_helpers import PERSON_UUID_IDENTIFIER_TYPE_ID
+
+        case_config_ids = case_config['patient_identifiers']
+        case_update = {}
+        id_type_uuid = PERSON_UUID_IDENTIFIER_TYPE_ID
+        if id_type_uuid in case_config_ids:
+            case_property = case_config_ids[id_type_uuid]['case_property']
+            value = patient['uuid']
+            case_update[case_property] = value
+        for identifier in patient['identifiers']:
```

Comment:
Are these the identifiers on the patient and now we are saving these on the case?
Would/Should these be blank before this or its fine to override them?

Reply:
Yes, they are OpenMRS's identifiers. `case_config_ids` are a subset of them that CommCare has been configured to store in case properties. 

We know that they are blank (or for some reason invalid), because if the case had valid values for at least one of them then the [`get_patient()` function](https://github.com/dimagi/commcare-hq/blob/nh/omrs/identifiers/corehq/motech/openmrs/repeater_helpers.py#L310) would have got the right patient before we had to fall back to searching. 

So this is why we're saving them to the case; so next time we need to match to an OpenMRS patient, we don't have to search; we can match straight away.